File: schedule-service/src/main/java/mrsisk/github/io/appointservice/service/UserService.java
package mrsisk.github.io.appointservice.service;

import mrsisk.github.io.appointservice.dto.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Service
public class UserService {

    @Autowired
    WebClient client;

    @Value("${accounts.base-uri}")
    String baseUrl;

    public Mono<User> getUserByEmail(String email){
       return client.get()
                .uri(baseUrl + "/auth/admin", c -> c.path("/users")
                        .queryParam("email", email)
                        .build()).exchangeToMono(clientResponse -> {
                    if (clientResponse.statusCode().is2xxSuccessful()){
                        return clientResponse.bodyToMono(User.class);
                    }
                    return Mono.empty();
                });
    }
}

File: accounts-service/src/main/kotlin/mrsisk/github/io/accountsmanager/service/UserService.kt
package mrsisk.github.io.accountsmanager.service

import mrsisk.github.io.accountsmanager.models.User
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.*
import org.springframework.web.reactive.function.server.*


@Service
class UserService(private val client: WebClient) {

    suspend fun findAllUsers(request: ServerRequest): ServerResponse{
        try {
          val users = client.get()
              .uri("/admin/realms/sisk/users")
              .awaitExchange {
                  if (it.statusCode().is2xxSuccessful) return@awaitExchange it.awaitBody(Array<User>::class)
                  else throw Exception("Error while fetching users")
              }
            return ServerResponse.ok().bodyValueAndAwait(users)
        }catch (err: Exception){
            return errorResponse(HttpStatus.NOT_FOUND, err.localizedMessage)
        }
    }

    suspend fun findUser(request: ServerRequest): ServerResponse{
        try {
            val id = request.pathVariable("id")
            val user = client.get()
                .uri("/admin/realms/sisk/users/$id")
                .awaitExchange {
                    if (it.statusCode().is2xxSuccessful) return@awaitExchange it.awaitBody(User::class)
                    else throw Exception("user $id Not found")
                }
            return ServerResponse.ok().json().bodyValueAndAwait(user)
        }catch (ex: Exception){
            return errorResponse(HttpStatus.NOT_FOUND, ex.localizedMessage)
        }


    }

    private suspend fun errorResponse(code: HttpStatus, message: String): ServerResponse{
        return ServerResponse.status(code).json().bodyValueAndAwait(mapOf("error" to message))
    }
}